backend/models/sequenceActivities.js

```diff
@@ -0,0 +1,298 @@
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	if (activity.parent && !await db.findOne(account,  activityCol(model), { _id: utils.stringToUUID(activity.parent)})) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const _id = nodeuuid();
+	activity = {...activity, sequenceId:  utils.stringToUUID(sequenceId), _id: utils.stringToUUID(_id)};
+
+	await Promise.all([
+		db.insert(account,  activityCol(model), activity),
+		FileRef.removeFile(account, model, "activities", sequenceId)
+	]);
+
+	return {...activity, _id, sequenceId: sequenceId};
+};
+
+SequenceActivities.edit = async (account, model, sequenceId, activityId, activity) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+
+	if (!activityEditSchema.isValidSync(activity, { strict: true })) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	if (activity.parent && !await db.findOne(account, activityCol(model), { _id: utils.stringToUUID(activity.parent)})) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const query =  {_id: utils.stringToUUID(activityId), sequenceId: utils.stringToUUID(sequenceId)};
+	const {result} = await db.update(account, activityCol(model), query, {$set: activity});
+
+	if (!result.n) {
+		throw responseCodes.ACTIVITY_NOT_FOUND;
+	}
+
+	await FileRef.removeFile(account, model, "activities", sequenceId);
+};
+
+SequenceActivities.remove = async (account, model, sequenceId, activityId) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+	const activities = await db.find(account,activityCol(model), {sequenceId: utils.stringToUUID(sequenceId), parent:{$exists: true}} , {parent:1, _id:1});
+
+	const idsToDelete = getDescendantsIds(activities, activityId);
+	idsToDelete.push(utils.stringToUUID(activityId));
+
+	// console.log(idsToDelete.map(utils.uuidToString));
+
+	const query = {_id:{ $in: idsToDelete}, sequenceId: utils.stringToUUID(sequenceId)};
+	const {result} = await db.remove(account,  activityCol(model), query);
+
+	if (!result.n) {
+		throw responseCodes.ACTIVITY_NOT_FOUND;
+	}
+
+	await FileRef.removeFile(account, model, "activities", sequenceId);
+
+	return { _id: activityId};
+};
+
+SequenceActivities.get = async (account, model, sequenceId) => {
+	const sequence = await db.findOne(account, model + ".sequences", { _id: utils.stringToUUID(sequenceId)});
+
+	if (!sequence) {
+		throw responseCodes.SEQUENCE_NOT_FOUND;
+	}
+
+	let activities = {};
+	try {
+		activities = await FileRef.getSequenceActivitiesFile(account, model, utils.uuidToString(sequenceId));
+	} catch(e) {
+		activities = await createActivitiesTree(account, model, sequenceId);
+		await FileRef.storeFile(account, activityCol(model) + ".ref", account, utils.uuidToString(nodeuuid()), JSON.stringify(activities),  {"_id": sequenceId});
```

Comment:
maybe we shouldn't await this one?

whether the cache is stored successfully into the fileshare or not doesn't affect the output of the network request. And we'd be waiting extra time when the user could use the data straight away

Reply:
What happens if the time between requests is less than the time it takes to store the file?